### ISSUE TITLE
TofuPipes

### ISSUE THREAD

**Issue by abstractionfactory** [opentofu/opentofu#2221]
This is a proposal from the community meeting today and aims to implement a subset of #931. The goal is to create an actionable featureset because the term "stacks" is overloaded and it is not clear that everyone agrees what it means.

## Goals

OpenTofu should gain a functionality where, based on a configuration file, it is able to access the state from another OpenTofu project located in a different directory and use the output values of that state as inputs for variables. The aim with this feature is to make it easier to refactor a single project into multiple projects and deduplicate a small part of the effort Terragrunt and many TACOS' have done.

## Non-goals

This proposal explicitly does not aim to solve problems that Terragrunt & co have solved, such as running multiple applies on multiple projects in parallel, etc.

## Technical considerations

Since OpenTofu is currently not usable as a library, this feature will probably rely on OpenTofu launching itself. See a lightweight prototype [here](https://github.com/opentofu/opentofu/pull/2517).

**Comment by apparentlymart** [comment]
I think this is tangentially related to what we were discussing over in https://github.com/opentofu/opentofu/issues/2160#issuecomment-2479758932: that idea introduces one possible place that someone could express the idea of reading some data from somewhere else and using it to populate an input variable:

```hcl
upstream "network" {
  # (some description of how to fetch some output values for "network")
}

variables = {
  environment = "staging"
  tier        = "data"

  subnet_ids = upstream.network.subnet_ids
}
```

In this example I'm imagining this new-style "workspace file" telling OpenTofu that it should, whenever preparing to take an action against this workspace, go fetch the output values for this upstream thing called "network" and then assign the upstream `subnet_ids` output value to this configuration's `subnet_ids` input variable.

Lots of other ways to do it of course... not meaning to suggest that this is definitely the right idea. In particular, that other issue is talking about totally revamping the idea of "workspaces" so it's a lot of baggage to attach to this idea. :grinning: Mainly this is just to create the cross-link between the issues.

**Comment by ImIOImI** [comment]
I'm inherently sceptical of sharing the remote state of a tofu stack with just any other project. This is, probably, a wrong headed approach, but I suggested the concept of a public/private way of accessing state info [in this issue here](https://github.com/opentofu/opentofu/issues/949). I however, fully believe we need a way to separate private and public concerns. I should be able to share the name of a key I use to store a password in AWS Key storage broadly without having to worry about accidentally leaking the actual key text string to the entire org. 

Furthermore, by separating public and private access we can better control RBAC access over whatever the bucket is that holds the state. For example, one team might not _need_ access to another team's stack's statefile, but it might be _very_ appropriate for them to want to share the name of a VPC they built.
